File: Ywy/JavaScript/level2/FunctionDevelopment.js
export default class FunctionDevelopment {
    solution = (progresses, speeds) => {
        const answer = [];
        const table = {};
        let count = 0;
        let max = 0;
        progresses.forEach((progress, index) => {
            let speed = speeds[index];
            let key = `${progress}_${speed}`;
            let day = table[key];
            if (!day) {
                day = this.getDay(progress, speed);
                if (max == 0) {
                    max = day;
                }
            }
            if (max < day) {
                answer.push(count);
                max = day;
                count = 0;
            }
            count++;
        });
        answer.push(count);
        return answer;
    };

    getDay = (progress, speed) => {
        let time = 0
        while(true) {
            time++;
            if (progress >= 100) {
                return time;
            }
            progress += speed;
        }
    };
}

File: Ywy/Go/level2/기능개발/functionDevelopment.go
package functionDevelopment

import "fmt"

func solution(progresses []int, speeds []int) []int {
	var (
		size                                                = len(progresses)
		answer                                              = make([]int, 0, size)
		table                                               = make(map[string]int, size)
		key                                                 string
		count, progress, speed, necessaryPeriod, max, index int
	)
	for {
		progress = progresses[index]
		speed = speeds[index]
		key = keyGen(progress, speed)
		necessaryPeriod = table[key]
		if necessaryPeriod == 0 {
			necessaryPeriod = getNecessaryPeriod(progress, speed)
			if max == 0 {
				max = necessaryPeriod
			}
		}
		if max < necessaryPeriod {
			answer = append(answer, count)
			count = 0
			max = necessaryPeriod
		}
		count++
		index++
		if index == size {
			answer = append(answer, count)
			return answer
		}
	}
}

func keyGen(progress, speed int) string {
	return fmt.Sprintf("%d_%d", progress, speed)
}

func getNecessaryPeriod(progress, speed int) (time int) {
	for {
		time++
		if progress >= 100 {
			return time
		}
		progress += speed
	}
}
